### ISSUE TITLE
`tonic::transport::Channel` struct is not as safe to clone for parallel use

### ISSUE THREAD

**Issue by AArnott** [hyperium/tonic#1751]
## Bug Report

It seems the `tonic::transport::Channel` struct is not as safe to clone as advertised. I use a static to store one channel after it is established, and then for every interested user, I return a clone of the Channel. But in doing this, my several tests almost always have one or more "transport error" failures. Simply removing the sharing by clone and forcing every client to establish their own novel channel totally fixes the tests, making them super reliable.

### Version

tonic 0.11.0
tonic-build v0.11.0

### Platform

x64 Windows
x64 Linux
x64 macOS

### Description

I have a multi-threaded application that makes a lot of gRPC calls, which tend to all go to just one or two servers. I understand that establishing a gRPC connection is expensive, so I'd rather reuse just one connection per server from many threads than to establish *n* connections, if that makes sense.

So as a very new rust programmer, I wrote the following `grpc.rs` file to make acquiring a `Channel` from any thread super-easy. 
The problem is that my tests reveal a great deal of instability. I *frequently* get test failures with "transport error" even though the server is _on the same machine_ as the gRPC client testing. Suspecting that the bug may be in the gRPC being shared for parallel use, I commented out the code in the file below related to _sharing_ of `Channel` values, instead giving each client its own new one. All the test failures went away. So surely, I'm doing something wrong. Can someone identify it?

```rs
use http::Uri;
use std::{cell::RefCell, collections::HashMap, sync::Mutex};
use tonic::transport::{Channel, ClientTlsConfig};
use zcash_client_backend::proto::service::compact_tx_streamer_client::CompactTxStreamerClient;

// We'll use a MUTEX to store the shareable gRPC channels, indexed by server URI.
// gRPC channels are expensive to create, cannot be used concurrently, but cheap to clone for each user.
lazy_static! {
    static ref CHANNELS: Mutex<HashMap<Uri, RefCell<Channel>>> = Mutex::new(HashMap::new());
}

/// Return a gRPC channel for the given URI, creating one if necessary.
pub(crate) async fn get_grpc_channel(uri: Uri) -> Result<Channel, tonic::transport::Error> {
    {
        let clients = CHANNELS.lock().unwrap();
        if let Some(client) = clients.get(&uri) {
            let channel = &*client.borrow();
            return Ok(channel.clone());
        }
    }

    let tls = ClientTlsConfig::new().domain_name(uri.host().unwrap());
    let channel = Channel::builder(uri.clone())
        .tls_config(tls)?
        .connect()
        .await?;

    let mut clients = CHANNELS.lock().unwrap();
    clients.insert(uri, RefCell::new(channel.clone()));
    Ok(channel)
}

/// Gets the CompactTxStreamerClient for the given URI for use with communicating with the lightwalletd server.
pub async fn get_client(
    uri: Uri,
) -> Result<CompactTxStreamerClient<Channel>, tonic::transport::Error> {
    let channel = get_grpc_channel(uri).await?;
    Ok(CompactTxStreamerClient::new(channel))
}
```

**Comment by AmeliasCode** [comment]
Does this only fail when running tests? And are you using the `tokio::test` attribute on your tests? IIRC there is a limitation with sharing statics across runtimes. By default, tokio will create a new runtime for each test.

**Comment by AArnott** [comment]
> Does this only fail when running tests?

I'm not sure. I haven't recorded any failure in normal application operation, but the 9 tests I have that all call into gRPC nearly at once is probably a more controlled concurrent environment than the random access used by the application itself.

Yes, I'm using `tokio::test` attributes on my tests. Maybe it's relevant that I notice a far better pass rate when using `cargo nextest` than `cargo test`.

> By default, tokio will create a new runtime for each test.

How does a unique runtime for each test impact this? Is the static value shared improperly in those conditions somehow?
I studied the issue you linked to and while I don't fully understand it, I applied one of the fixes proposed by the community and it appears to have worked: https://github.com/nerdcash/Nerdbank.Cryptocurrencies/pull/352

I'd love to understand it better though.
